docs/tutorial.md

```diff
@@ -32,13 +32,122 @@ require('vendor/autoload.php');
 
-## Doing parallel requests
+```php
+use Http\Discovery\HttpAsyncClientDiscovery;
 
-TODO explain sendRequests and how to work with BatchResult and BatchException
+$httpAsyncClient = HttpAsyncClientDiscovery::find();
+$promise = $httpAsyncClient->sendAsyncRequest($request);
+```
+
+This promise allows you to : 
+
+ * Add callbacks for when the response is available or an errors happens by using the then method:
+ 
+ ```php
+ $promise->then(function (ResponseInterface $response) {
+     // onFulfilled callback
+     echo 'The response is available';
+    
+     return $response;
+ }, function (Exception $e) {
+     // onRejected callback
+     echo 'An error happens';
+    
+     throw $e;    
+ });
+ ```
+ 
+ This method will return another promise so you can manipulate the response and/or exception and
+ still provide a way to interact with this object for your users:
+ 
+ ```php
+ $promise->then(function (ResponseInterface $response) {
+     // onFulfilled callback
+     echo 'The response is available';
+     
+     return $response;
+ }, function (Exception $e) {
+     // onRejected callback
+     echo 'An error happens';
+     
+     throw $e;    
+ })->then(function (ResponseInterface $response) {
+     echo 'Response stil available';
+    
+     return $response;
+ }, function (Exception $e) {
+      throw $e
+ });
+ ```
+  
+ In order to achieve the chain callback, if you read previous examples carefully, callbacks provided to the `then` method __must__ 
+ return a PSR-7 `ResponseInterface` or throw a `Http\Client\Exception`. For both of the callbacks, if it returns a PSR-7 `ResponseInterface` 
+ it will call the `onFulfilled` callback for the next element in the chain, if it throws a `Http\Client\Exception` it will call the `onRejected`
+ callback.
+ 
+ i.e. you can inverse the behavior of a call:
+ 
+ ```php
+  $promise->then(function (ResponseInterface $response) use($request) {
+      // onFulfilled callback
+      echo 'The response is available, but it\'s not ok...';
+      
+      throw new HttpException('My error message', $request, $response);
+  }, function (Exception $e) {
+      // onRejected callback
+      echo 'An error happens, but it\'s ok...';
+      
+      return $exception->getResponse();  
+  });
+  ```
+  
+ * Get the state of the promise with `$promise->getState()` will return of one `Promise::PENDING`, `Promise::FULFILLED` or `Promise::REJECTED`
+ * Get the response of the promise if it's in `FULFILLED` state with `$promise->getResponse()` call
+ * Get the error of the promise if it's in `REJECTED` state with `$promise->getRequest()` call
+ * wait for the callback to be fulfilled or rejected with the `$promise->wait()` call. The `wait` will return nothing, it will simply call one the callback
+ pass to the `then` method depending on the result of the call. It the promise has already been fulfilled or rejected it will do nothing.
```

Comment:
i think this is too much detail for the tutorial. the tutorial should show a typical basic use case and can link to the reference documentation for the full list of options. otherwise the tutorial will get very large, and contain information that is not found elsewhere - which would be unexpected.